torch/csrc/distributed/rpc/rref.h

```diff
@@ -0,0 +1,137 @@
+#pragma once
+
+#include <c10/util/Optional.h>
+#include <torch/csrc/distributed/rpc/message.h>
+#include <torch/csrc/distributed/rpc/rpc_agent.h>
+#include <torch/csrc/distributed/rpc/types.h>
+
+#include <atomic>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+
+class RRef;
+class RRefContext;
+class UserRRef;
+
+// Represents fork of an RRef to be sent over the wire.
+//
+// In order to preserve correctness of reference counting, each RRefForkData
+// **MUST** be deserialized into a RRef. This means that if RRefForkData is to
+// be transferred across the network, we need the guarantee that the message
+// will *eventually* get to the peer,  and that the peer will create a RRef out
+// of it. Therefore, no constructor of RRefForkData is exposed, and
+// applications should never directly use RRefForkData. All construction are
+// done within ``RRef`` and ``RRefContext``.
+struct RRefForkData {
+  at::IValue toIValue() const;
+ private:
+  friend class RRef;
+  friend class RRefContext;
+  friend class UserRRef;
+
+  RRefForkData(worker_id_t ownerId,
+               const RRefId& rrefId_,
+               const ForkId& forkId_);
+
+  static RRefForkData fromIValue(at::IValue&&);
+
+  const worker_id_t ownerId_;
+  const RRefId rrefId_;
+  const ForkId forkId_;
+};
```

Comment:
If this is sent over the wire as a buffer, without going through serialization, then it pays off to add:

```cpp
static_assert(std::is_trivially_copyable<RRefForkData>::value, "RRefForkData must be trivially copyable");
```

Reply:
It seems we need GCC version >= 5 to use `is_trivially_copyable`?

Reply:
Hmm crap... You're right.

But we have the following macro instead: https://github.com/pytorch/pytorch/blob/d8314a6260dd8c6de1d8e32d18e138d2b2d64cc1/c10/macros/Macros.h#L213-L217

Reply:
Let me use that.